### PR TITLE
[x/TWAP] Expose a geometric TWAP API 

### PR DESCRIPTION
**Pull request by stackman27**
Closes: Part of https://github.com/osmosis-labs/osmosis/issues/3514 [1-3]

## What is the purpose of the change
We need to expose a geometric TWAP API. Its implementation would be almost identical to the arithmetic TWAP API.
Therefore, we must develop proper abstractions to reduce code duplication and avoid complexity.

## Brief Changelog
n/a

## Testing and Verifying
n/a

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:10]: Let's add godocs please for each of:
- `twapStrategy`
- `getTwapToNow`
- `computeTwap`

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:33]: What do you mean by "decide between arithmetic or geometric" here? I think this might confuse folks.

`arithmetic` strategy should always utilize the respective arithmetic methods

File: x/twap/api.go
Comment (p0mvn) [inline x/twap/api.go:71]: let's add godoc please explaining that this is a more generalized method taking a strategy. Based on the strategy taken, it would execute different logic

File: x/twap/api.go
Comment (p0mvn) [inline x/twap/api.go:99]: Same godoc suggestion

File: x/twap/strategy.go
Comment (stackman27) [reply x/twap/strategy.go:33]: isn't the idea to have a `isArithmeticType` bool to decide if we want to calculate arithmetic or geometric in this function?

**Comment by stackman27** [comment]
@p0mvn Added some simple godocs because i couldn't figure out a lot to add. lmk what you think

File: x/twap/strategy.go
Comment (p0mvn) [reply x/twap/strategy.go]: There is no `isArithmeticType` boolean in this function.

If you are referring to some changes introduced in: https://github.com/osmosis-labs/osmosis/pull/3420, those would actually need to be refactored to be integrated with the strategy pattern here. So the boolean parameter `isArithmeticTwap` in #3420 would likely be changed to taking a strategy instead.

I suggest assuming that #3420 does not exist for the context of this PR. We will integrate them after one or the other is merged

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:11]:
```suggestion
// twapStrategy is an interface for computing TWAPs.
// We have two strategies implementing the interface - arithmetic and geometric.
// We expose a common TWAP API to reduce duplication and avoid complexity.
```

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:28]: Let's move this to be defined within the interface declarion, please.

The reason is that when you hover over a method in your editor if there is a godoc above the interface, it will be shown

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:40]: Let's also move this to be defined within the interface pls

```suggestion
// computeTwap calculates the TWAP with specific startRecord and endRecord.
```

File: x/twap/api.go
Comment (p0mvn) [inline x/twap/api.go:102]:
```suggestion
// getTwapToNow computes and returns twap from the start time until the current block time. The type
// of twap returned depends on the strategy given and can be either arithmetic or geometric.
```

File: x/twap/api.go
Comment (p0mvn) [inline x/twap/api.go:72]:
```suggestion
// getTwap computes and returns twap from the start time until the end time. The type
// of twap returned depends on the strategy given and can be either arithmetic or geometric.
```

File: x/twap/api.go
Comment (p0mvn) [inline x/twap/api.go:50]: Generally, we should treat a function as a black box. Using a spec, we aim to convey what we get by calling the function given certain inputs.

Since twap record is an internal implementation detail that clients should not be concerned about from the black box perspective, I suggest rephrasing it

```suggestion
// GetArithmeticTwapToNow returns arithmetic twap from start time until the current block time for quote and base 
// assets in a given pool.
```

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:36]: nit: From our previous discussion, I'm not sure I'm following the suggestion in the comment, and I'm afraid that this might confuse other readers as well. I suggest removing it unless there is something I'm missing. 
```suggestion
```

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:42]: ditto
```suggestion
```

File: x/twap/api.go
Comment (p0mvn) [inline x/twap/api.go:63]:
```suggestion
// GetBeginBlockAccumulatorRecord returns a TwapRecord struct corresponding to the state of pool `poolId`
// and given assets as of the beginning of the block this is called on.
```

File: x/twap/api.go
Comment (p0mvn) [inline x/twap/api.go:66]: I'm not sure I'm following everything described here. I think the method is as simple as returning a most recent record with accumulator values interpolated to the current block time.

**Comment by p0mvn** [comment]
> @p0mvn Added some simple godocs because i couldn't figure out a lot to add. lmk what you think

Added some suggestions. LGTM once addressed!

File: x/twap/api.go
Comment (stackman27) [reply x/twap/api.go:66]: hmm, not sure about the nuances of this function as this existed before this PR. I can remove it if u suggest

**Comment by stackman27** [comment]
Resolved comments [`bcad2b2`](https://github.com/osmosis-labs/osmosis/pull/3529/commits/bcad2b25e1fcf6b48a3cbb8010c737ccfb6d6515)

File: x/twap/api.go
Comment (p0mvn) [reply x/twap/api.go:66]: I suggest removing it because it felt misleading to me. I think we should focus on the most basic functionality of what the function does in its spec

**Review by p0mvn — APPROVED** [review]
LGTM. Nice work

File: x/twap/strategy.go
Comment (stackman27) [reply x/twap/strategy.go]: aah gotchu

File: x/twap/api.go
Comment (stackman27) [reply x/twap/api.go:66]: gotchu will do

**Comment by stackman27** [comment]
@p0mvn rebased and good to go

**Comment by p0mvn** [comment]
@czarcas7ic @nicolaslara hey, can we get one more review here please?

**Review by czarcas7ic — APPROVED** [review]
This LGTM

File: x/twap/strategy.go
Comment (nicolaslara) [inline x/twap/strategy.go:22]: nit: Should we unify these names: either both be called "get" or both called "compute"?

Comment (nicolaslara) [reply x/twap/strategy.go:22]: also. Is there a reason to have `getTwapToNow` be part of the strategy?

**Review by nicolaslara — APPROVED** [review]
Some nits in the comments, but other than that this LGTM

File: x/twap/strategy.go
Comment (stackman27) [reply x/twap/strategy.go:22]: i think unifying makes sense, i'm up for naming it `getTwapToNow` and `getTwap`. @p0mvn lmk what you think of this

For `getTwapToNow` I was calling the keeper method directly [here](https://github.com/osmosis-labs/osmosis/pull/3529/files#diff-4d165167a6ae9c6e7d42cffc254bb4a98eb339ccf26be3a032a40ca2729a5ce8R87) but i changed it to use strategy.go route

File: x/twap/api.go
Comment (p0mvn) [inline x/twap/api.go:87]: Just realized something from @nicolaslara 's [comment](https://github.com/osmosis-labs/osmosis/pull/3529/files#r1035678186).

The only reason we define `getTwapToNow` on strategy is to be able to call it here.

However, we can instead do the following:
```go
k.getTwapToNow(ctx, poolId, baseAssetDenom, quoteAssetDenom, startTime, strategy)
```

and remove the `getTwapToNow` from strategy in general

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:20]: Please see my other comment.

```suggestion
```

File: x/twap/strategy.go
Comment (p0mvn) [inline x/twap/strategy.go:39]:
```suggestion
```

File: x/twap/api.go
Comment (stackman27) [reply x/twap/api.go:87]: sounds good

File: x/twap/api.go
Comment (stackman27) [reply x/twap/api.go:126]: Should we update https://github.com/osmosis-labs/osmosis/issues/3514 to add it there then?

**Review by p0mvn — APPROVED** [review]

**Pull request merged**
